traitsui/qt4/extra/styled_tree_editor.py

```diff
@@ -0,0 +1,192 @@
+                nid, node, instance, index=index
+            )
+
+        if index is None:
+            cnid = QtGui.QTreeWidgetItem(nid)
+        else:
+            cnid = QtGui.QTreeWidgetItem()
+            nid.insertChild(index, cnid)
+
+        cnid.setToolTip(0, node.get_tooltip(instance))
+        self._set_column_labels(cnid, node.get_column_labels(instance))
+
+        color = node.get_background(instance)
+        if color:
+            cnid.setBackground(0, self._get_brush(color))
+
+        color = node.get_foreground(instance)
+        if color:
+            cnid.setForeground(0, self._get_brush(color))
+
+        return cnid
+
+    def _update_icon(self, nid):
+        """ Override to avoid (again) calls to `QTreeWidgetItem.setIcon`
+        """
+        expanded, node, instance = self._get_node_data(nid)
+        if not node.delegates_style:
+            nid.setIcon(0, self._get_icon(node, instance, expanded))
+
+
+class StyledTreeEditor(TreeEditorFactory):
+    """ The EditorFactory for a StyledTreeEditor
+    """
+    #: A StyledTreeItemDelegate subclass which handles the drawing
+    item_delegate_klass = Type
+
+    def _get_simple_editor_class(self):
+        return SimpleEditor
+
+
+class StyledTreeItemDelegate(QtGui.QStyledItemDelegate):
+    """ An ItemDelegate base class which handles `sizeHint`
+    """
+    def __init__(self, *args, **kwargs):
+        self.size_map = defaultdict(lambda: QtCore.QSize(1, 21))
+        QtGui.QStyledItemDelegate.__init__(self, *args, **kwargs)
+
+    def custom_paint(self, painter, option, index, node, instance):
+        """ The paint() method for subclasses to override.
+        """
+        raise NotImplementedError()
+
+    def paint(self, painter, option, index):
+        """ Draw the item """
+        # Qt gives us some drawing for free
+        super(StyledTreeItemDelegate, self).paint(painter, option, index)
+
+        item = self.editor._tree.itemFromIndex(index)
+        _, node, instance = self.editor._get_node_data(item)
+
+        if node.delegates_style:
+            self.custom_paint(painter, option, index, node, instance)
+        else:
+            # Need to set the appropriate sizeHint of the item.
+            self.updateSizeHint(index, option.rect.size())
+
+    def sizeHint(self, option, index):
+        """ Returns area taken by the item. """
+        return self.size_map[self.editor._tree.itemFromIndex(index)]
+
+    def updateSizeHint(self, index, size):
+        """ Updates the size stored in the size_map. """
+        item = self.editor._tree.itemFromIndex(index)
+        if self.size_map[item] != size:
+            self.size_map[item] = size
+            do_later(self.sizeHintChanged.emit, index)
+
+
+class StyledTreeNode(TreeNode):
+    """ A TreeNode to use with StyledTreeEditor
+    """
+    #: Set to True to use ItemDelegate for drawing
+    delegates_style = Bool(False)
+
+
+if __name__ == '__main__':
```

Comment:
Could you put this into the `examples` directories (not sure where is best, since it is Qt-specific.

Reply:
That shouldn't be a problem